src/matrix.hpp

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include "spaces.hpp"
+
+namespace opt {
+
+template <Vector V, std::size_t Cols>
+struct matrix {
+    static constexpr std::size_t rows = std::tuple_size<V>::value;
+    static constexpr std::size_t cols = Cols;
+    using entries_type = scalar_t<V>;
+    using columns_type = V;
+
+    using row_index = typename V::coords_type::size_type;
+    using col_index = std::size_t;
+
+    std::array<V, Cols> columns{};
+
+    // Special member functions defaulted
+    constexpr matrix() = default;
+
+    // Ctor from elements, row major
+    template <class... Us>  // clang-format off
+        requires(
+            (std::same_as<entries_type, std::remove_cvref_t<Us>> && ...) &&
+            (sizeof...(Us) == (cols * rows)))
+    constexpr matrix(Us&&... us)  // clang-format on
+    {
+        row_index row{0};
+        col_index col{0};
+        (
+            [&row, &col, &the_columns = columns](auto&& elem) {
+                // FIXME use index sequences
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
+                the_columns[col][row] = elem;
+                col = (col == cols - 1) ? 0 : col + 1;
+                row = (col == 0) ? row + 1 : row;
+            }(std::forward<Us>(us)),
+            ...);
+    }
+
+    // Ctor from Vectors
+    template <class... Us>  // clang-format off
+        requires(
+            (std::same_as<columns_type, std::remove_cvref_t<Us>> && ...) &&
```

Comment:
This could use some sort of utility metafunction. Not sure of a good name though

Reply:
I am not at all convinced matrices should be constructed from "n Vectors", actually. For example dense matrices are stored in a totally different way. Either now or as soon as the equality stuff is in I think it would be nicer if there was a matrix concept - basically a double indexable vector, in some sense. I am bad at abstraction so I don't know the details though :)

Reply:
Anyway, back to your comment. Do you mean something like this?
template<class Target, class... Us>
concept all_essentially_same = (std::same_as<Target, std::remove_cvref_t<Us>> && ...);